robosuite/utils/mjmod.py

```diff
@@ -0,0 +1,935 @@
+    #     # for collisions.
+    #     self._geom_ids_by_tex_id = defaultdict(list)
+    #     for geom_id in range(self.model.ngeom):
+    #         mat_id = self.model.geom_matid[geom_id]
+    #         if mat_id >= 0:
+    #             tex_id = self.model.mat_texid[mat_id]
+    #             if tex_id >= 0:
+    #                 self._geom_ids_by_tex_id[tex_id].append(geom_id)
+
+    def _cache_checker_matrices(self):
+        """
+        Cache two matrices of the form [[1, 0, 1, ...],
+                                        [0, 1, 0, ...],
+                                        ...]
+        and                            [[0, 1, 0, ...],
+                                        [1, 0, 1, ...],
+                                        ...]
+        for each texture. To use for fast creation of checkerboard patterns
+        """
+        self._texture_checker_mats = []
+        for tex_id in range(self.model.ntex):
+            texture = self.textures[tex_id]
+            h, w = texture.bitmap.shape[:2]
+            self._texture_checker_mats.append(self._make_checker_matrices(h, w))
+
+    def _make_checker_matrices(self, h, w):
+        re = np.r_[((w + 1) // 2) * [0, 1]]
+        ro = np.r_[((w + 1) // 2) * [1, 0]]
+        cbd1 = np.expand_dims(np.row_stack(((h + 1) // 2) * [re, ro]), -1)[:h, :w]
+        cbd2 = np.expand_dims(np.row_stack(((h + 1) // 2) * [ro, re]), -1)[:h, :w]
+        return cbd1, cbd2
+
+
+# From mjtTexture
+MJT_TEXTURE_ENUM = ['2d', 'cube', 'skybox']
+
+class Texture():
+    """
+    Helper class for operating on the MuJoCo textures.
+    """
+
+    __slots__ = ['id', 'type', 'height', 'width', 'tex_adr', 'tex_rgb']
+
+    def __init__(self, model, tex_id):
+        self.id = tex_id
+        self.type = MJT_TEXTURE_ENUM[model.tex_type[tex_id]]
+        self.height = model.tex_height[tex_id]
+        self.width = model.tex_width[tex_id]
+        self.tex_adr = model.tex_adr[tex_id]
+        self.tex_rgb = model.tex_rgb
+
+    @property
+    def bitmap(self):
+        size = self.height * self.width * 3
+        data = self.tex_rgb[self.tex_adr:self.tex_adr + size]
+        return data.reshape((self.height, self.width, 3))
+
+class PhysicalParameterModder:
+    """
+    Modder for various physical parameters of the mujoco model
+    can use to modify parameters stored in MjModel (ie friction, damping, etc.) as
+    well as optimizer parameters like global friction multipliers (eg solimp, solref, etc)
+    To modify a parameteter, use the parameter to be changed as a keyword argument to
+    self.mod and the new value as the value for that argument. Supports arbitray many
+    modifications in a single step.
+    NOTE: It is necesary to perform sim.forward after performing the modification.
+    NOTE: Some parameters might not be able to be changed. users are to verify that
+          after the call to forward that the parameter is indeed changed.
+    """
+    def __init__(self, sim):
+        self.sim = sim
```

Comment:
should this extend from `BaseModder`?

Reply:
This is copied from @aprendizaje-de-maquinas - I did not test this yet. I'm not going to handle dynamics randomization in this PR - we can merge that in a separate PR once we decide how to handle it and what kinds to do.